designdocs/user-handles.md

```diff
@@ -0,0 +1,308 @@
+    App->>RPC: get_msa_id_for_handle(handle)
+    RPC->>Chain: Query handle mapping state (State Query MSA<->Handle)
+    Chain-->>RPC: Return handle mapping state (State Query MSA<->Handle)
+    RPC-->>App: Return MSA ID for handle
+    User->>App: Request to get handle for MSA ID
+    App->>RPC: get_handle_for_msa_id(msa_id)
+    RPC->>Chain: Query MSA ID mapping state (Reverse Index Handle<->MSA)
+    Chain-->>RPC: Return MSA ID mapping state (Reverse Index Handle<->MSA)
+    RPC-->>App: Return handle for MSA ID
+```
+
+## Storage
+
+* **HandleSuffixMin**: This storage value (```u32``) will keep track of the minimum suffix value allowed for a given handle.
+* **HandleSuffixMax**: This storage value (```u32``) will keep track of the maximum suffix value allowed for a given handle.
+* **HandleRetirementPeriod**: This storage value (```u32``) will keep track of the number of blocks a handle must be retired before it can be reused.
+* **MSAIdToUserHandle**: This storage will map MSAIds to their corresponding user handles.
+* **UserHandlesToMSAId**: This storage will map user handles to their corresponding MSAIds.
+* **RetiredHandles**: This storage will keep track of all retired handles and the block number at which they were retired. This probably is a spike to run few ideas on how to implement this.
+* **UsedSuffixes**: This storage will keep track of all used suffixes to ensure that no two handles have the same suffix. (May not be necessary)
+* **Seed (current and previous)**: This storage values (```tuple (s0, s1)```) will keep track of the current and previous seed values.
+
+## Required Extrinsics
+
+### Primitives
+
+``` rust
+
+MsaHandlePayload {
+    handle: &[u8],
+    suffix: u32,
+    msa_id: u64,
+}
+```
+
+### Create user handle with chosen suffix
+
+ As a network frequency should allow users to choose their own handle and suffix. This extrinsic will allow users to create a handle with a chosen suffix.
+
+``` rust
+Input
+
+* origin - must be a signed origin
+* owner msa id - the MSA ID of the user. MSA ID must be created before calling this extrinsic.
+* proof - the proof of ownership of the handle, ```MsaHandlePayload``` signed by the user/provider private key
+
+Output
+* Event - `MsaHandleCreated` with the MSA ID and the handle
+
+* Errors - 
+    * `HandleAlreadyExists` if the handle already exists
+    * `InvalidMsaHandle` if the handle is invalid
+    * `Unauthorized` if the signature is invalid
+    * `InvalidSuffix` if the suffix is invalid
+    * `SuffixAlreadyExists` if the suffix already exists
+
+Validation requirements
+* handle must follow handle guidelines.
+* suffix must be in the range of the handle.
+* handle must be unique.
+* handle must be available.
+* suffix must be available.
+* MSA ID must exist.
+
+Signature requirements
+
+The extrinsic must be signed by the user/provider private key. The signature must be verified on-chain to ensure that the user is the owner of the private key.
```

Comment:
Side note, but a reminder that as this does not have a general replay issue, it does NOT need to be added into the signature registry to protect against replay.

Reply:
Totally 👍